src/SensirionShdlcCommunication.cpp

```diff
@@ -124,15 +124,18 @@ uint16_t SensirionShdlcCommunication::receiveFrame(SensirionShdlcRxFrame& frame,
         return RxFrameError | BufferSizeError;
     }
 
-    size_t i = 0;
-    while (i < dataLength) {
-        error = unstuffByte(current, serial, startTime, timeoutMicros);
-        if (error) {
-            return error;
+    // An error frame does not have any data eventhough the value of "dataLength" is not zero.
+    // Therefore, only read data from the buffer if the frame state is 0, i.e, the frame is not of an error frame.
```

Comment:
If the frame state is `!= 0` the MISO frame contains one byte with the execution error code. Since we always read the data length the execution error code is read into the dataLength variable. 
Later on line 156 we are handling this error.

The correct thing to do, in my opinion, is to handle the error after the state us read out. So if the status is not 0 we should read the error code and set the dataLength to 0. This way everything should be handled correctly and we still read out the whole frame and don't need to handle the state == 0 here.

Reply:
True that the system would return an error when an error frame is arriving, however, this never actually happened. Instead, when the error frame came, the system got stuck for a long while, then it restarted.  This was because it got stuck at the loop reading data from the unstuffByte and it continued reading even through the end of the buffer was reach, and so the arduino restarted.

So, in order to fix this issue and make sure that the error frame is treated as an error frame and that there is no way to read anything beyond the end of the buffer, I added the if case to only read data if the frame is normal.

I guess there could be other ways to solve this issue. unstuffByte could return an error if it encounters a 0x7e, but that would maybe make it harder to see what is going on....

Reply:
yes in case of an error it should not read any additional data.
This is why I suggested to set the dataLength to 0 in case of an error frame which has the same effect as your current solution.

We probably don't want to exit the function early because there will be still a CRC and end frame in case of an error frame.

What I suggest is (untested code) is something like this
```
if (frame._state == 0) {
    error = unstuffByte(dataLength, serial, startTime, timeoutMicros);
    if (error)
        return error;
} else {
    // error frame was received
    uint8_t errorCode;
    dataLength = 0;
    error = unstuffByte(errorCode, serial, startTime, timeoutMicros);
    if (error)
        return error;
}
```

If the dataLength is 0 it will not get stuck in the loop later on, as there is no further data to read.

Reply:
Yes, the whole frame needs to be read, as it is not an time out. 
Yes, your solution would work as well. I have no preference.